torch/fx/subgraph_rewriter.py

```diff
@@ -83,7 +85,114 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
+                # use?
+                else:
+                    raise RuntimeError("Attempted to replace ",
+                                       node.target, " during subgraph "
+                                       "rewriting, but both the "
+                                       "original GraphModule `gm` and "
+                                       "the replacement nn.Module "
+                                       "`replacement` have a submodule"
+                                       " of this name. The two "
+                                       "submodules reference different"
+                                       " objects, so it is unclear "
+                                       "what ", node.target, " should"
+                                       " refer to")
+
+            # CASE 3: The target was part of the original GraphModule
+            # `gm`, but it wasn't a submodule
+            elif not gm_submod and gm_err == ErrType.EXISTS_AS_OTHER_ATTR:
+                raise RuntimeError("Attempted to replace ", node.target,
+                                   " during subgraph rewriting, but ",
+                                   node.target, " already exists in the"
+                                   " original GraphModule `gm` as "
+                                   "something other than an nn.Module")
+
+            # CASE 4: The target wasn't part of `gm`, but we found the
+            # correct replacement submodule in `replacement`
+            elif (not gm_submod and gm_err == ErrType.DOES_NOT_EXIST 
+                    and replacement_submod):
+                new_submod = copy.deepcopy(getattr(replacement, node.target))
+                gm.add_submodule(node.target, new_submod)
+
+            # CASE 5: The target wasn't part of `gm`, and the
+            # replacement we found wasn't a submodule
+            elif (not gm_submod and not replacement_submod
+                    and replacement_err == ErrType.EXISTS_AS_OTHER_ATTR):
+                raise RuntimeError("Attempted to replace ", node.target,
+                                   " during subgraph rewriting, but ",
+                                   node.target, " does not exist in the"
+                                   " original GraphModule `gm`. The "
+                                   "candidate replacement submodule "
+                                   "in `replacement` is not an "
+                                   "nn.Module")
+
+            # CASE 5: The target wasn't part of `gm` or `replacement`
+            else:
+                raise RuntimeError("Attempted to replace ", node.target,
+                                   " during subgraph rewriting, but ",
+                                   node.target, " does not exist in the"
+                                   " original GraphModule `gm` or in "
+                                   "the replacement nn.Module "
+                                   "`replacement`")
```

Comment:
When trying to understand this code I drew out this diagram:

```
                        gm has module
                      +---------------------+----------------+
                      | False               |    True        |
              +-------+---------------------+----------------+
Replacement   | False | ERROR               +  use gm mod    |
              |       |                     |  w/ type check |
has module    +-------+---------------------+----------------+
              | True  | copy w/ type check  | use gm mod     |
              +-------+---------------------+----------------+
```

Some thoughts I have are:

1) The `gm`'s submodule takes precedence over the replacement's submodule
2) Copy over the submodule from replacement only if `gm` does not have that target

I think the above two conditions could be encoded pretty simply, and the only error condition would be if neither graphmodule has the referenced target submodule.

3) I think type checking should be done elsewhere, e.g. we could simply do checks for if `call_module` truly is calling into a module in `lint`: https://github.com/pytorch/pytorch/blob/dd2bccafc599b3931555d95c222809a94d1cbe28/torch/fx/graph.py#L1010. Then we could just call `lint` after we trace the replacement callable and call lint after we're done with subgraph rewriting